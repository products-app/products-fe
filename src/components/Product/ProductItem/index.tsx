import { Text, css } from '@lebernardo/react'
import { formatDecimalToReal, getProductImage } from '@/helpers/products'
import {
  ProductItemContainer,
  ButtonContainer,
  ButtonAddToCart,
  ProductName,
} from './styles'
import Image from 'next/image'
import { useProduct } from '../hook'

type ProductItemProps = {
  product: app.Product
  css?: typeof css
}

const ProductItem = ({ product, css }: ProductItemProps) => {
  const { handleAddToCart, isDisabledAddToCart } = useProduct()

  return (
    <ProductItemContainer css={{ css }}>
      <Image
        src={getProductImage(product.image)}
        alt={product.name}
        width={0}
        height={0}
        sizes="100vw"
      />

      <figcaption>
        <ProductName size="sm">{product.name}</ProductName>
        <Text>{formatDecimalToReal(product.price)}</Text>

        <ButtonContainer>
          <ButtonAddToCart
            variant="outline"
            onClick={() => handleAddToCart(product)}
            disabled={isDisabledAddToCart(product)}
          >
            Adicionar ao carrinho
          </ButtonAddToCart>
        </ButtonContainer>
      </figcaption>
    </ProductItemContainer>
  )
}

const defaultProps = {
  css: '',
}

ProductItem.defaultProps = defaultProps
export default ProductItem
