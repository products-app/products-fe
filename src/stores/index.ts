import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'
import { createUserSlice } from './slices/userSlice'
import { createCartSlice } from './slices/cartSlice'
import { createAppSlice } from './slices/appSlice'

export const useStore = create(
  persist<app.ICart & app.IUser & app.IAppControl>(
    (set, get) => ({
      ...createUserSlice(set),
      ...createCartSlice(set, get),
      ...createAppSlice(set),
    }),
    {
      name: 'store',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
