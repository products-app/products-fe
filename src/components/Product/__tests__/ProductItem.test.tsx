import { render, screen, renderHook } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ProductItem from '../ProductItem'
import { useStore } from '@/stores'

const productMock: app.Product = {
  id: 1,
  name: 'Product Test',
  price: 29.9,
  stock: 2,
}

describe('Product Item component', () => {
  it('should render correctly', () => {
    const { container } = render(<ProductItem product={productMock} />)

    expect(screen.getByText('Product Test')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should click to add product in the cart store', async () => {
    render(<ProductItem product={productMock} />)

    const handleClick = jest.fn()
    const buttonElement = screen.getByRole('button', {
      name: 'Adicionar ao carrinho',
    })
    buttonElement.addEventListener('click', handleClick, false)
    await userEvent.click(buttonElement)

    expect(handleClick).toHaveBeenCalled()

    const { result } = renderHook(() => useStore())
    const cartItems = result.current.items
    expect(cartItems?.length).toEqual(1)
  })
})
