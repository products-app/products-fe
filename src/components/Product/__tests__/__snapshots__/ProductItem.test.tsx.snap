// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Item component should render correctly 1`] = `
<div>
  <div
    class="c-jYQuMM c-jYQuMM-ijdaZMY-css"
  >
    <img
      alt="Product Test"
      data-nimg="1"
      decoding="async"
      height="0"
      loading="lazy"
      sizes="100vw"
      src="/_next/image?url=%2Fproduct-not-found.jpg&w=3840&q=75"
      srcset="/_next/image?url=%2Fproduct-not-found.jpg&w=640&q=75 640w, /_next/image?url=%2Fproduct-not-found.jpg&w=750&q=75 750w, /_next/image?url=%2Fproduct-not-found.jpg&w=828&q=75 828w, /_next/image?url=%2Fproduct-not-found.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproduct-not-found.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproduct-not-found.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproduct-not-found.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproduct-not-found.jpg&w=3840&q=75 3840w"
      style="color: transparent;"
      width="0"
    />
    <figcaption>
      <h2
        class="c-dgMvMW c-hZwmvA c-dgMvMW-iPGoSV-size-sm"
      >
        Product Test
      </h2>
      <p
        class="c-dgMvMW c-dgMvMW-kWyXas-size-md"
      >
        R$Â 29,90
      </p>
      <div
        class="c-eOfRZm"
      >
        <button
          class="c-gJqLDS c-hihFAR c-gJqLDS-esnCTh-variant-outline c-gJqLDS-bIqJQp-size-medium"
        >
          Adicionar ao carrinho
        </button>
      </div>
    </figcaption>
  </div>
</div>
`;
